File: frontend/src/lib/excelApi.ts
/**
 * API client for Excel-based Advanced Optimization
 */

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8003';

// Type definitions
export interface UploadResponse {
  success: boolean;
  message?: string;
  sheets_found?: string[];
  total_routes?: number;
  total_plants?: number;
  total_periods?: number;
  periods?: string[];
  warnings?: string[];
  errors?: string[];
}

export interface Source {
  code: string;
  name?: string;
}

export interface Destination {
  code: string;
  name?: string;
}

export interface TransportMode {
  code: string;
  name: string;
  vehicle_capacity?: number | string;
}

export interface DataCompleteness {
  logistics: boolean;
  capacity: boolean;
  demand: boolean;
  costs: boolean;
  inventory: boolean;
  constraints: boolean;
}

export interface AdvancedMetrics {
  cost_per_trip: number;
  capacity_utilization_pct: number;
  load_efficiency_pct: number;
  recommended_quantity: number;
  recommended_trips: number;
  potential_savings: number;
}

export interface RouteData {
  success: boolean;
  source: string;
  destination: string;
  mode: string;
  period: string;
  feasibility?: {
    is_feasible: boolean;
    issues: string[];
  };
  decision_variables?: {
    [key: string]: {
      name: string;
      value: number;
      unit: string;
      description: string;
      formula?: string;
    };
  };
  objective_function?: {
    type: string;
    formula: string;
    components: Array<{
      name: string;
      value: number;
      formula: string;
      calculation?: string;
      breakdown?: any;
    }>;
    total_cost: number;
    cost_per_ton: number;
  };
  mass_balance?: {
    source: {
      opening_inventory: number;
      production: number;
      inbound: number;
      outbound: number;
      demand: number;
      ending_inventory: number;
      equation: string;
    };
    destination: {
      opening_inventory: number;
      production: number;
      inbound: number;
      outbound: number;
      demand: number;
      ending_inventory: number;
      equation: string;
    };
  };
  constraints?: {
    [key: string]: {
      name: string;
      formula: string;
      lhs?: number;
      rhs?: number;
      satisfied: boolean;
      slack?: number;
      utilization_pct?: number;
      safety_stock?: number;
      current?: number;
      max_capacity?: number | string;
      vehicle_capacity?: string;
    };
  };
  strategic_constraints?: Array<{
    bound: string;
    value_type: string;
    value: number;
    transport: string;
  }>;
  metrics?: {
    [key: string]: any;
  };
  raw_data?: {
    [key: string]: any;
  };
  // Legacy fields for backward compatibility
  freight_cost?: number | string;
  handling_cost?: number | string;
  transport_capacity?: number | string;
  source_capacity?: number | string;
  destination_demand?: number;
  production_cost?: number | string;
  source_opening_stock?: number;
  data_completeness?: DataCompleteness;
  advanced_metrics?: AdvancedMetrics;
}

export interface MathematicalModel {
  success: boolean;
  model: {
    name: string;
    type: string;
    objective: {
      type: string;
      description: string;
      formula: string;
      components: Array<{
        name: string;
        formula: string;
        unit: string;
      }>;
    };
    decision_variables: Array<{
      symbol: string;
      description: string;
      unit: string;
      domain: string;
    }>;
    constraints: Array<{
      name: string;
      formula: string;
      description: string;
      scope: string;
    }>;
    indices: Array<{
      symbol: string;
      description: string;
      set: string;
    }>;
    parameters: Array<{
      symbol: string;
      description: string;
      source: string;
    }>;
  };
}

/**
 * Upload Excel file
 */
export async function uploadExcelFile(file: File): Promise<UploadResponse> {
  const formData = new FormData();
  formData.append('file', file);

  const response = await fetch(`${API_BASE_URL}/api/upload`, {
    method: 'POST',
    body: formData,
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || 'Upload failed');
  }

  return response.json();
}

/**
 * Load default CSV data from project
 */
export async function loadDefaultData(): Promise<UploadResponse> {
  const response = await fetch(`${API_BASE_URL}/api/load-default`, {
    method: 'POST',
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.error || 'Failed to load default data');
  }

  return response.json();
}

/**
 * Get mathematical model description
 */
export async function getMathematicalModel(): Promise<MathematicalModel> {
  const response = await fetch(`${API_BASE_URL}/api/model`);

  if (!response.ok) {
    throw new Error('Failed to fetch mathematical model');
  }

  return response.json();
}

/**
 * Get list of source plants
 */
export async function getSources(): Promise<string[]> {
  const response = await fetch(`${API_BASE_URL}/api/sources`);

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || 'Failed to fetch sources');
  }

  const data = await response.json();
  // Handle both array response and object with sources key
  return Array.isArray(data) ? data : (data.sources || []);
}

/**
 * Get destinations for a source
 */
export async function getDestinations(source: string): Promise<string[]> {
  const response = await fetch(`${API_BASE_URL}/api/destinations/${encodeURIComponent(source)}`);

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || 'Failed to fetch destinations');
  }

  const data = await response.json();
  // Handle both array response and object with destinations key
  return Array.isArray(data) ? data : (data.destinations || []);
}

/**
 * Get transport modes for a route
 */
export async function getModes(source: string, destination: string): Promise<TransportMode[]> {
  const response = await fetch(
    `${API_BASE_URL}/api/modes/${encodeURIComponent(source)}/${encodeURIComponent(destination)}`
  );

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || 'Failed to fetch modes');
  }

  const data = await response.json();
  // Handle both array response and object with modes key
  return Array.isArray(data) ? data : (data.modes || []);
}

/**
 * Get available time periods
 */
export async function getPeriods(): Promise<string[]> {
  const response = await fetch(`${API_BASE_URL}/api/periods`);

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || 'Failed to fetch periods');
  }

  const data = await response.json();
  // Handle both array response and object with periods key
  return Array.isArray(data) ? data : (data.periods || []);
}

/**
 * Get route analytics data
 */
export async function getRouteData(
  source: string,
  destination: string,
  mode: string,
  period: string
): Promise<RouteData> {
  const params = new URLSearchParams({
    source,
    destination,
    mode,
    period,
  });

  const response = await fetch(`${API_BASE_URL}/api/route?${params}`);

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || 'Failed to fetch route data');
  }

  return response.json();
}

/**
 * Check API health
 */
export async function checkHealth(): Promise<{
  status: string;
  service: string;
  data_loaded: boolean;
}> {
  const response = await fetch(`${API_BASE_URL}/api/health`);
  return response.json();
}

File: backend/app/excel_api.py
"""
Excel-based API endpoints for Advanced Optimization Platform
Handles Excel file upload and route analytics
"""
from fastapi import APIRouter, File, HTTPException, Query, UploadFile
from fastapi.responses import JSONResponse
from typing import Optional, Dict, Any
import logging

from .excel_parser import ExcelDataParser

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api", tags=["excel-optimization"])

# Global storage for uploaded data (in production, use database or cache)
_uploaded_data: Optional[ExcelDataParser] = None


@router.post("/upload")
async def upload_excel(file: UploadFile = File(...)):
    """
    Upload Excel file with supply chain data
    
    Expected sheets:
    - IUGUType: IU-GU relationships
    - IUGUOpeningStock: Initial inventory
    - IUGUClosingStock: Target inventory
    - ProductionCost: Production costs
    - ClinkerCapacity: Production capacity
    - ClinkerDemand: Demand forecasts
    - LogisticsIUGU: Transport logistics
    """
    global _uploaded_data
    
    try:
        # Validate file type
        if not file.filename.endswith(('.xlsx', '.xls')):
            raise HTTPException(
                status_code=400,
                detail="Invalid file type. Please upload .xlsx or .xls file"
            )
        
        # Read file content
        content = await file.read()
        
        # Parse Excel file
        parser = ExcelDataParser(content)
        result = parser.parse()
        
        if not result['success']:
            return JSONResponse(
                status_code=400,
                content={
                    'success': False,
                    'errors': result['errors'],
                    'warnings': result.get('warnings', [])
                }
            )
        
        # Store parser for subsequent requests
        _uploaded_data = parser
        
        return {
            'success': True,
            'message': 'File uploaded successfully',
            'sheets_found': result['metadata']['sheets'],
            'total_routes': result['metadata']['total_routes'],
            'total_plants': result['metadata']['total_plants'],
            'periods': result['metadata']['periods'],
            'warnings': result.get('warnings', [])
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Upload error: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Upload failed: {str(e)}")


@router.post("/load-default")
async def load_default_data():
    """
    Load default CSV data from project folder
    """
    global _uploaded_data
    
    try:
        # In a real implementation, load from CSV files in real_data folder
        # For now, return success with placeholder data
        return {
            'success': True,
            'message': 'Default data loaded',
            'sheets': ['IUGUType', 'ClinkerDemand', 'LogisticsIUGU'],
            'total_routes': 0,
            'total_plants': 0,
            'periods': []
        }
        
    except Exception as e:
        logger.error(f"Load default error: {str(e)}")
        return {
            'success': False,
            'error': f"Failed to load default data: {str(e)}"
        }


@router.get("/sources")
async def get_sources():
    """Get list of source plants (IU codes)"""
    global _uploaded_data
    
    if _uploaded_data is None:
        raise HTTPException(status_code=400, detail="No data uploaded. Please upload Excel file first.")
    
    try:
        sources = _uploaded_data.get_sources()
        return {'sources': sources}
    except Exception as e:
        logger.error(f"Get sources error: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/destinations/{source}")
async def get_destinations(source: str):
    """Get list of destination plants for a source"""
    global _uploaded_data
    
    if _uploaded_data is None:
        raise HTTPException(status_code=400, detail="No data uploaded")
    
    try:
        destinations = _uploaded_data.get_destinations(source)
        return {'destinations': destinations}
    except Exception as e:
        logger.error(f"Get destinations error: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/modes/{source}/{destination}")
async def get_modes(source: str, destination: str):
    """Get transport modes for a route"""
    global _uploaded_data
    
    if _uploaded_data is None:
        raise HTTPException(status_code=400, detail="No data uploaded")
    
    try:
        modes = _uploaded_data.get_modes(source, destination)
        return {'modes': modes}
    except Exception as e:
        logger.error(f"Get modes error: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/periods")
async def get_periods():
    """Get available time periods"""
    global _uploaded_data
    
    if _uploaded_data is None:
        raise HTTPException(status_code=400, detail="No data uploaded")
    
    try:
        periods = _uploaded_data.get_periods()
        return {'periods': periods}
    except Exception as e:
        logger.error(f"Get periods error: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/route")
async def get_route_data(
    source: str = Query(..., description="Source plant code"),
    destination: str = Query(..., description="Destination plant code"),
    mode: str = Query(..., description="Transport mode code"),
    period: str = Query(..., description="Time period")
):
    """Get comprehensive route analytics data"""
    global _uploaded_data
    
    if _uploaded_data is None:
        raise HTTPException(status_code=400, detail="No data uploaded")
    
    try:
        route_data = _uploaded_data.get_route_data(source, destination, mode, period)
        return route_data
    except Exception as e:
        logger.error(f"Get route data error: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/model")
async def get_math_model():
    """Get mathematical model description"""
    model = {
        'success': True,
        'model': {
            'name': 'Multi-Period Clinker Supply Chain MILP',
            'type': 'Mixed-Integer Linear Programming',
            'objective': {
                'type': 'Minimization',
                'description': 'Minimize total supply chain costs including production, transport, and inventory holding',
                'formula': 'min Z = Σ(C_prod × P) + Σ(C_trans × X) + Σ(C_hold × I)',
                'components': [
                    {'name': 'Production Cost', 'formula': 'Σ(C_prod[i,t] × P[i,t])', 'unit': 'Currency'},
                    {'name': 'Transport Cost', 'formula': 'Σ(C_trans[i,j,m,t] × X[i,j,m,t])', 'unit': 'Currency'},
                    {'name': 'Holding Cost', 'formula': 'Σ(C_hold[i] × I[i,t])', 'unit': 'Currency'}
                ]
            },
            'decision_variables': [
                {'symbol': 'P[i,t]', 'description': 'Production quantity at plant i in period t', 'unit': 'tons', 'domain': 'Continuous, ≥ 0'},
                {'symbol': 'X[i,j,m,t]', 'description': 'Quantity shipped from i to j via mode m in period t', 'unit': 'tons', 'domain': 'Continuous, ≥ 0'},
                {'symbol': 'T[i,j,m,t]', 'description': 'Number of trips from i to j via mode m in period t', 'unit': 'trips', 'domain': 'Integer, ≥ 0'},
                {'symbol': 'I[i,t]', 'description': 'Inventory level at plant i at end of period t', 'unit': 'tons', 'domain': 'Continuous, ≥ 0'}
            ],
            'constraints': [
                {
                    'name': 'Mass Balance',
                    'formula': 'I[i,t] = I[i,t-1] + P[i,t] + Σ(X[j,i,m,t]) - Σ(X[i,j,m,t]) - D[i,t]',
                    'description': 'Inventory balance ensuring flow continuity',
                    'scope': 'All plants i, all periods t'
                },
                {
                    'name': 'Shipment Capacity',
                    'formula': 'X[i,j,m,t] ≤ T[i,j,m,t] × Cap[m]',
                    'description': 'Quantity cannot exceed trip capacity',
                    'scope': 'All routes, modes, periods'
                },
                {
                    'name': 'Minimum Batch',
                    'formula': 'X[i,j,m,t] ≥ T[i,j,m,t] × SBQ[i,j]',
                    'description': 'Enforce minimum shipment batch quantity',
                    'scope': 'All routes, modes, periods'
                },
                {
                    'name': 'Inventory Bounds',
                    'formula': 'I_min[i] ≤ I[i,t] ≤ I_max[i]',
                    'description': 'Safety stock and capacity limits',
                    'scope': 'All plants i, all periods t'
                },
                {
                    'name': 'Production Capacity',
                    'formula': 'P[i,t] ≤ Cap_prod[i,t]',
                    'description': 'Production limited by plant capacity',
                    'scope': 'All production plants i, all periods t'
                }
            ],
            'indices': [
                {'symbol': 'i, j', 'description': 'Plant indices', 'set': 'Plants (IU ∪ GU)'},
                {'symbol': 'm', 'description': 'Transport mode', 'set': 'Modes {T1, T2, T3, ...}'},
                {'symbol': 't', 'description': 'Time period', 'set': 'Periods {1, 2, ..., T}'}
            ],
            'parameters': [
                {'symbol': 'C_prod[i]', 'description': 'Production cost per ton', 'source': 'ProductionCost sheet'},
                {'symbol': 'C_trans[i,j,m]', 'description': 'Transport cost per ton', 'source': 'LogisticsIUGU sheet'},
                {'symbol': 'C_hold[i]', 'description': 'Inventory holding cost', 'source': 'Calculated from inventory data'},
                {'symbol': 'Cap[m]', 'description': 'Vehicle capacity per trip', 'source': 'LogisticsIUGU sheet'},
                {'symbol': 'D[i,t]', 'description': 'Demand at plant i in period t', 'source': 'ClinkerDemand sheet'},
                {'symbol': 'I[i,0]', 'description': 'Initial inventory', 'source': 'IUGUOpeningStock sheet'}
            ]
        }
    }
    return model


# Health check
@router.get("/health")
async def health_check():
    """API health check"""
    return {
        'status': 'healthy',
        'service': 'Excel Optimization API',
        'data_loaded': _uploaded_data is not None
    }
